### PR TITLE
Support for uploading notebooks

### PR DESCRIPTION
**Pull request by aniezurawski**

### CODE REVIEW

File: neptune/projects.py [hunk 1]
```diff
@@ -281,6 +282,20 @@ def create_experiment(self,
 
         return experiment
 
+    def create_notebook(self, file_path):
+        if not file_path.endswith(".ipynb"):
```
Comment (hubertjaworskineptune) [inline neptune/projects.py]: This validation seems common and could be moved to a util function

File: neptune/client.py [hunk 2]
```diff
@@ -667,9 +713,14 @@ def _upload_loop_chunk(self, fun, part, data, **kwargs):
...
-        url = self.api_address + api_method.operation.path_name
-        url = url.replace("{experimentId}", experiment.internal_id)
+    def _upload_raw_data(self, api_method, data, headers, path_params, query_params):
+        url = self.api_address + api_method.operation.path_name + "?"
+
+        for key, val in path_params.iteritems():
+            url = url.replace("{" + key + "}", val)
+
+        for key, val in query_params.iteritems():
+            url = url + key + "=" + val + "&"
```
Comment (hubertjaworskineptune) [inline neptune/client.py]: this will produce a query ending with '&'

Comment (aniezurawski) [reply neptune/client.py]: Yeah, and it's totally fine.

File: neptune/client.py [hunk 3]
```diff
@@ -269,6 +272,80 @@ def create_experiment(self,
...
+                "jupyterPath": jupyter_path
+            }
+        )
+
+        if response.status_code == NOT_FOUND:
+            raise NotebookNotFound(notebook_id=notebook_id)
+        elif response.status_code != OK:
+            raise RuntimeError("Http status code {}: {}".format(response.status_code, response.reason))
+
+        response.close()
```
Comment (lukasz-walkiewicz) [inline neptune/client.py]: You should close resources in the finally block or use a context manager.

File: neptune/client.py [hunk 4]
```diff
@@ -322,20 +399,23 @@ def update_tags(self, experiment, tags_to_add, tags_to_delete):
...
-                raise StorageLimitReached()
-            else:
-                raise
+        elif (response.status_code == UNPROCESSABLE_ENTITY
+              and extract_response_field(response.content, 'type') == 'LIMIT_OF_STORAGE_IN_PROJECT_REACHED'):
+            raise StorageLimitReached()
+        elif response.status_code != OK:
+            raise RuntimeError("Http status code {}: {}".format(response.status_code, response.reason))
+
+        response.close()
```
Comment (lukasz-walkiewicz) [inline neptune/client.py]: Same as above

File: neptune/client.py [hunk 5]
```diff
@@ -269,6 +272,80 @@ def create_experiment(self,
...
+            },
+            query_params={
+                "jupyterPath": jupyter_path
+            }
+        )
+
+        if response.status_code == NOT_FOUND:
+            raise NotebookNotFound(notebook_id=notebook_id)
+        elif response.status_code != OK:
+            raise RuntimeError("Http status code {}: {}".format(response.status_code, response.reason))
```
Comment (lukasz-walkiewicz) [inline neptune/client.py]: Create a new exception type for this kind of errors.

File: neptune/client.py [hunk 6]
```diff
@@ -703,9 +787,14 @@ def _upload_loop_chunk(self, fun, part, data, **kwargs):
                    },
                    **kwargs)
 
-    def _upload_raw_data(self, experiment, api_method, data, headers):
-        url = self.api_address + api_method.operation.path_name
-        url = url.replace("{experimentId}", experiment.internal_id)
+    def _upload_raw_data(self, api_method, data, headers, path_params, query_params):
```
Comment (lukasz-walkiewicz) [inline neptune/client.py]: Shouldn't content-type header be set here?

File: neptune/notebook.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        self._project = project
+        self._id = _id
+        self._owner = owner
+
+    @property
+    def id(self):
+        """ Notebook id
+
+        Examples:
+            Instantiate a session and fetch a project.
```
Comment (lukasz-walkiewicz) [inline neptune/notebook.py]: Is it necessary to duplicate the project initialization part of the comment?

File: neptune/utils.py [hunk 8]
```diff
@@ -45,6 +45,17 @@ def as_list(value):
         return [value]
 
 
+def validate_notebook_path(path):
+    if not path.endswith(".ipynb"):
+        raise RuntimeError("'{}' is not a correct notebook file. Should end with '.ipynb'.".format(path))
```
Comment (lukasz-walkiewicz) [inline neptune/utils.py]: Add a new exception type for these errors

File: neptune/client.py [hunk 9]
```diff
@@ -269,6 +272,80 @@ def create_experiment(self,
...
+                project=project,
+                _id=api_notebook.id,
+                owner=api_notebook.owner
+            )
+        except HTTPNotFound:
+            raise NotebookNotFound(notebook_id=notebook_id, project=project.full_id)
+
+    def get_last_checkpoint(self, project, notebook_id):
+        try:
+            api_checkpoint_list = self.leaderboard_swagger_client.api.listCheckpoints(
```
[the remark this answers is not in the document]
Comment (aniezurawski) [reply neptune/client.py]: Yes. Latest first.

File: neptune/client.py (hunk 5 above)
Comment (aniezurawski) [reply neptune/client.py]: This is fixed in next pull request.

File: neptune/client.py (hunk 3 above)
Comment (aniezurawski) [reply neptune/client.py]: This is fixed in next pull request.

File: neptune/client.py (hunk 6 above)
Comment (aniezurawski) [reply neptune/client.py]: Caller of this method is responsible for setting all necessary headers.

**Pull request merged**